python/activator/logger.py

```diff
@@ -171,17 +171,62 @@ def __init__(self, labels=None):
             self._labels = {}
 
     def format(self, record):
-        # Call for side effects only; ignore result.
-        super().format(record)
+        # format updates record.message, but the full info is *only* in the return value.
+        msg = super().format(record)
 
         entry = {
             "severity": record.levelname,
             "logging.googleapis.com/labels": self._labels | record.logging_context,
-            "message": record.getMessage(),
+            "message": msg,
         }
         return json.dumps(entry)
 
 
+class UsdfJsonFormatter(logging.Formatter):
+    """A formatter that can be parsed by the Loki/Grafana system at USDF.
+
+    The formatter's output is a JSON-encoded message with "flattened" metadata
+    to make it easy to inspect with Grafana filters.
+
+    Parameters
+    ----------
+    labels : `dict` [`str`, `str`]
+        Any metadata that should be attached to the log.
+    """
+    def __init__(self, labels=None):
+        super().__init__()
+
+        if labels:
+            self._labels = labels
+        else:
+            self._labels = {}
+
+    def format(self, record):
+        # format updates record.message, but the full info is *only* in the return value.
+        msg = super().format(record)
+
+        # Many LogRecord attributes are only useful for interrogating the
+        # record in Python; filter to what's useful in Grafana.
+        entry = {
+            # formatTime only automatically handles msecs with the default format???
```

Comment:
Is it intentional to leave the `???` ?

Reply:
You're right, that was rude of me. I'll reword.